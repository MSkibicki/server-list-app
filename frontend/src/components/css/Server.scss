@import "./Global.scss";

.server {
  background-color: $white;
  border: 2px solid $list-border-color;
  list-style: none;
  display: flex;
  justify-content: space-around;
  align-items: center;
  margin: 0 1px;

  h1 {
    font-size: $server-name-font;
    font-weight: 600;
    width: 40%;
  }

  h3 {
    font-size: $server-status-font;
    font-weight: 600;
  }

  h3.online {
    color: $online-color;
  }

  h3.offline {
    color: $offline-color;
  }
}
