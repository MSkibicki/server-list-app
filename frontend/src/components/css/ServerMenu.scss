@import "./Global.scss";

.server-icon:hover {
  cursor: pointer;
}
