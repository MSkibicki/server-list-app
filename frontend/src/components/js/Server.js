import React, { useState } from "react";
import "../css/Server.scss";
import ServerMenu from "./ServerMenu";
import axios from "axios";
import classNames from "classnames";

const Server = ({ name, status, id }) => {
  const [serverStatus, setServerStatus] = useState(null);

  const serverOn = () => {
    axios
      .put(`http://localhost:5000/servers/${id}/on`, { status: "ONLINE" })
      .then((response) => {
        setServerStatus(response.data.status);
      });
  };

  const serverOff = () => {
    axios
      .put(`http://localhost:5000/servers/${id}/off`, { status: "OFFLINE" })
      .then((response) => {
        setServerStatus(response.data.status);
      });
  };

  const serverClass = classNames({
    "online": status === "ONLINE",
    "offline": status === "OFFLINE",
    "put-online": serverStatus !== null && serverStatus === "ONLINE",
    "put-offline": serverStatus !== null && serverStatus === "OFFLINE",
  });

  return (
    <li className="server">
      <h1>{name}</h1>
      <h3 className={serverClass}>{serverStatus ? serverStatus : status}</h3>
      <ServerMenu
        serverOn={serverOn}
        serverOff={serverOff}
        status={status}
        serverStatus={serverStatus}
      />
    </li>
  );
};

export default Server;
