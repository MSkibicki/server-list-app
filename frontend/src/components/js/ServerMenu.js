import React, { useState } from "react";
import "../css/ServerMenu.scss";
import MoreHorizIcon from "@material-ui/icons/MoreHoriz";
import Menu from "@material-ui/core/Menu";
import MenuItem from "@material-ui/core/MenuItem";

const ServerMenu = ({ serverOn, serverOff, serverStatus }) => {
  const [anchorEl, setAnchorEl] = useState(null);

  const handleClick = (event) => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = () => {
    setAnchorEl(null);
  };

  const handleOnOff = () => {
    handleClose();
    if (serverStatus === "ONLINE") {
      serverOff();
    } else {
      serverOn();
    }
  };

  return (
    <div className="server-icon">
      <MoreHorizIcon
        aria-controls="simple-menu"
        aria-haspopup="true"
        onClick={handleClick}
      ></MoreHorizIcon>
      <Menu
        id="simple-menu"
        anchorEl={anchorEl}
        keepMounted
        open={Boolean(anchorEl)}
        onClose={handleClose}
      >
        <MenuItem onClick={handleOnOff}>
          {`
          ${serverStatus === "ONLINE" ? "Turn Off" : "Turn On"}
          `}
        </MenuItem>
        <MenuItem onClick={handleClose}>Reboot</MenuItem>
      </Menu>
    </div>
  );
};

export default ServerMenu;
