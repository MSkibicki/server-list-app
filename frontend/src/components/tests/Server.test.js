import Server from "./../js/Server";
import ServerMenu from "./../js/ServerMenu";

describe("Server />", () => {
  it("renders correctly", () => {
    const wrapper = shallow(<Server />);

    expect(wrapper).toMatchSnapshot();
  });

  it("has correct li classname", () => {
    const wrapper = shallow(<Server />);

    const liClass = wrapper.find("li.server").hasClass("server");

    expect(liClass).toBe(true);
  });

  it("renders <ServerMenu /> component", () => {
    const wrapper = shallow(<Server />);

    const menu = wrapper.find(ServerMenu);

    expect(menu.length).toEqual(1);
  });
});
