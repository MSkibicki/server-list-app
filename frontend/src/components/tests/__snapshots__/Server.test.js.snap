// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server /> renders correctly 1`] = `
<li
  className="server"
>
  <h1 />
  <h3
    className=""
  />
  <ServerMenu
    serverOff={[Function]}
    serverOn={[Function]}
    serverReboot={[Function]}
  />
</li>
`;
